### PR TITLE
[mcache] Switch to python-binary-memcached & support SASL authentication

### PR DESCRIPTION
**Pull request by dirk**
Enables connecting to Memcached servers with SASL username/password authentication via the binary protocol. Non-authenticated connections are still supported.

**dd-agent-omnibus PR**: https://github.com/DataDog/dd-agent-omnibus/pull/80

### CODE REVIEW

**Comment by dirk** [comment]
This addresses issue #1333.

**Comment by degemer** [comment]
Hey @dirk, thanks for the PR!
We are now going to decide whether we will review this PR for our 5.9 release or a future one.
Please ping me if I haven't updated this PR at the end of next week.

**Comment by dirk** [comment]
@degemer: Thanks Quentin! We'd very much like to have this capability for monitoring our Memcached instances at Everlane, so I have time available to put towards any changes/improvements y'all want.

**Comment by degemer** [comment]
@dirk: it will be part of 5.9, someone will review it in the coming weeks.

You already opened PR for dd-agent and dd-agent-omnibus, could you also add one for omnibus-software (https://github.com/DataDog/dd-agent/blob/master/CONTRIBUTING.md#omnibus-software-1) ?

**Comment by dirk** [comment]
@degemer: Done! See DataDog/omnibus-software#61.

File: checks.d/mcache.py [hunk 1]
```diff
@@ -162,18 +163,18 @@ class Memcache(AgentCheck):
...
     def get_library_versions(self):
-        return {"memcache": memcache.__version__}
+        return {
+            "memcache": pkg_resources.get_distribution("python-binary-memcached").version
+        }
 
     def _get_metrics(self, client, tags, service_check_tags=None):
         try:
-            raw_stats = client.get_stats()
+            stats_by_host = client.stats()
```
Comment (truthbk) [inline checks.d/mcache.py]: Took a look at the reference docs (http://python-binary-memcached.readthedocs.io/en/latest/bmemcached/#module-bmemcached.client), `stats()` will return a dict, and `stats(key=None)` suggests the parameter would imply grabbing the stats for a certain key.... 

Please correct me if I'm wrong, but wouldn't we expect this call to return stats for all keys if `key is None` and thus having multiple keys the `len(stats_by_host)` could be greater than 1?

**Comment by truthbk** [comment]
Some tests mocking the `client.stats()` response would be fantastic and would provide some peace of mind (being `memcached` such an important integration). I can try to provide a meaningful output we could use in the mock.

**Comment by dirk** [comment]
@truthbk: Are you describing unit tests of the implementation based on mocked output of the memcached API's `stats` call? If so, I can get some sample output and put that together. As far as peace-of-mind goes, there are already—thankfully!—fairly thorough integration tests in [`tests/checks/integration/test_mcache.py`](https://github.com/DataDog/dd-agent/blob/74f674645b64ac47f51d94cd847a0d562414dbf0/tests/checks/integration/test_mcache.py) which are [passing on CI](https://travis-ci.org/DataDog/dd-agent/jobs/180153349).

File: checks.d/mcache.py (hunk 1 above)
Comment (dirk) [inline checks.d/mcache.py]: @truthbk: The reason I have the 1-length assertion is that the Memcached check is configured to only talk to one server (per [`mcache.yaml`](https://github.com/DataDog/dd-agent/blob/74f674645b64ac47f51d94cd847a0d562414dbf0/conf.d/mcache.yaml.example)), so getting a 0-sized dict would indicate the server didn't report stats and a greater-than-1-sized dict would indicate multiple servers reporting stats/the single server misreporting stats. Please let me know if you think a different approach should be taken! 🙃

**Comment by masci** [comment]
Closing as superseded by https://github.com/DataDog/integrations-core/pull/1364

**Pull request closed without merging**